server/src/main/java/org/apache/druid/metadata/SegmentsMetadataManagerConfig.java

```diff
@@ -31,6 +31,9 @@ public class SegmentsMetadataManagerConfig
   @JsonProperty
   private Period pollDuration = new Period("PT1M");
 
+  @JsonProperty
+  private boolean useFindFullyOvershadowed;
```

Comment:
We shouldn't need a config to choose a better code path. If the new flow is better, it should be used in all cases.

If we need some config to avoid using the new flow in cases where segment locking has been used, then the config should disable segment locking itself and throw an exception if someone tries to use segment locking. Only then can we be sure that the coordinator is operating safely.

Reply:
The new flow is empirically better for large clusters with thousands of segments per interval.
There may not be much of a benefit for smaller clusters and I had added the config to allow a rollback to the current flow since the methods in the new flow are not well documented.

An alternative that I was thinking of was to detect if an interval has segment-locking enabled i.e if there are segments with distinct minor versions. If not, we optimize the existing isOvershadowed call to work faster with timechunk-locking based segments. Does this seem like a better approach?

Reply:
I've tried out the alternative with benchmarks. It's simpler and seems very effective.
Could you please share your thoughts?

Reply:
It does make sense to check if an interval has any segments with a minor version before falling back to the old logic. How costly is it to find such segments? Is it only an O(n) operation?

Reply:
Yes, it is.
I think there may be a flaw in the current method though. (It assumes that the interval is immutable)
Perhaps we could just maintain the maxValue in the holder for an interval which is updated whenever new segments are added.

Reply:
> It assumes that the interval is immutable

I think this is true for the places where `isOvershadowed` is being called currently, but it is easy to ensure that it works without such an assumption as well

Reply:
I don't understand the meaning of "Interval is immutable". When can interval be mutable?

Reply:
When a new segment is added to or removed from the interval

Reply:
Ah, I see, you mean whether you would have to update the value of `maxMinorVersion` during the lifecycle of a `TimelineEntry`?

Reply:
Yes

Reply:
Thanks, Instead of maintaining a map that is updated with the count when add or remove are called, I've kept a simple max of the minor version which is incremented when a segment with a greater minor version is added.

Reply:
That makes sense. 👍🏻